crates/engine/tree/src/tree/mod.rs

```diff
@@ -2650,6 +2663,245 @@ where
+                .unwrap_or_else(|| block.parent_num_hash());
+
+            self.state.buffer.insert_block(block);
+
+            return Ok(InsertPayloadOk::Inserted(BlockStatus::Disconnected {
+                head: self.state.tree_state.current_canonical_head,
+                missing_ancestor,
+            }))
+        };
+
+        // now validate against the parent
+        let parent_block = self.sealed_header_by_hash(block.parent_hash())?.ok_or_else(|| {
+            InsertBlockErrorKind::Provider(ProviderError::HeaderNotFound(
+                block.parent_hash().into(),
+            ))
+        })?;
+        if let Err(e) =
+            self.consensus.validate_header_against_parent(block.sealed_header(), &parent_block)
+        {
+            warn!(target: "engine::tree", ?block, "Failed to validate header {} against parent: {e}", block.hash());
+            return Err(e.into())
+        }
+
+        // We only run the parallel state root if we are currently persisting blocks that are all
+        // ancestors of the one we are executing. If we're committing ancestor blocks, then: any
+        // trie updates being committed are a subset of the in-memory trie updates collected before
+        // fetching reverts. So any diff in reverts (pre vs post commit) is already covered by the
+        // in-memory trie updates we collect in `compute_state_root_parallel`.
+        //
+        // See https://github.com/paradigmxyz/reth/issues/12688 for more details
+        let is_descendant_of_persisting_blocks =
+            self.is_descendant_of_persisting_blocks(block.header());
+
+        let Some(provider_builder) = self.state_provider_builder(block.parent_hash())? else {
+            // TODO(mattsse): this is the same logic as the `state_provider` call above and should
+            // be unified
+            unreachable!()
+        };
+
+        // use prewarming background task
+        let header = block.clone_sealed_header();
+        let txs = block.clone_transactions_recovered().collect();
+        let mut handle = if is_descendant_of_persisting_blocks && self.config.use_state_root_task()
+        {
+            // use background tasks for state root calc
+            let consistent_view = ConsistentDbView::new_with_latest_tip(self.provider.clone())?;
+
+            // Compute trie input
+            let trie_input_start = Instant::now();
+            let trie_input = self
+                .compute_trie_input(consistent_view.clone(), block.header().parent_hash())
+                .map_err(|e| InsertBlockErrorKind::Other(Box::new(e)))?;
+
+            self.metrics
+                .block_validation
+                .trie_input_duration
+                .set(trie_input_start.elapsed().as_secs_f64());
+
+            self.payload_processor.spawn(header, txs, provider_builder, consistent_view, trie_input)
+        } else {
+            self.payload_processor.spawn_prewarming(header, txs, provider_builder)
```

Comment:
this makes sense, but feels a bit weird... Maybe we can move spawning the prewarming outside the if clause, and pass the prewarming handle into `payload_processor.spawn` if the state root task is enabled?